dask/array/core.py

```diff
@@ -1109,12 +1109,20 @@ def itemsize(self):
         """ Length of one array element in bytes """
         return self.dtype.itemsize
 
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, val):
+        self._name = val
+        # Clear the key cache when the name is reset
+        self._cached_keys = None
+
     def _keys(self, *args):
         if not args:
-            try:
+            if self._cached_keys is not None:
```

Comment:
> Do you know __slots__ affects dictionary key sharing?

I'm not sure I follow how that's relevant here/I don't see `__slots__` in that pep. Can you explain?

> Maybe now that you check of self._cached_keys is not None you can add self._cached_keys = None to the __init__.

We assign to `self.name` in `__init__`, which hits the setter and sets `self._cached_keys = None` already. Keeps the cache clearing all in one spot.